File: jvm/src/main/kotlin/com/jetbrains/signatureverifier/serialization/TaggedObjectInfo.kt
package com.jetbrains.signatureverifier.serialization

import kotlinx.serialization.Serializable
import org.bouncycastle.asn1.ASN1Encodable
import org.bouncycastle.asn1.ASN1Primitive
import org.bouncycastle.asn1.DLTaggedObject

@Serializable
data class TaggedObjectInfo(
  val explicit: Boolean,
  val tagNo: Int,
  val content: EncodableInfo
) : EncodableInfo {
  companion object {
    private const val DECLARED_EXPLICIT = 1
    private const val DECLARED_IMPLICIT = 2
    private const val PARSED_EXPLICIT = 3
    private const val PARSED_IMPLICIT = 4

    /**
     * Hack to get same explicitness as in original
     */
    fun getTaggedObject(
      explicit: Boolean,
      tagNo: Int,
      content: ASN1Encodable
    ): DLTaggedObject = DLTaggedObject(explicit, tagNo, content)
  }

  override fun toPrimitive(): ASN1Primitive = getTaggedObject(explicit, tagNo, content.toPrimitive()).toASN1Primitive()

}

File: net/JetBrains.SignatureVerifier/src/Serialization/TaggedObjectInfo.cs
using Org.BouncyCastle.Asn1;

namespace JetBrains.SignatureVerifier.Serialization;

public class TaggedObjectInfo : IEncodableInfo
{
  private readonly bool _explicit;
  private readonly int _tagNo;
  private readonly IEncodableInfo _content;
  
  public TaggedObjectInfo(bool explicitness, int tagNo, IEncodableInfo content)
  {
    _explicit = explicitness;
    _tagNo = tagNo;
    _content = content;
  }

  public static DerTaggedObject GetTaggedObject(bool explicitness, int tagNo, Asn1Encodable content)
  {
    return new DerTaggedObject(explicitness, tagNo, content);
  }

  public Asn1Encodable ToPrimitive()
  {
    return GetTaggedObject(_explicit, _tagNo, _content.ToPrimitive()).ToAsn1Object();
  }
}